clifton/CoopGame
# add game mode

### set up eqs logic

- re-derive `BP_TestGameMode` to newly created `SGameMode`
![image](https://user-images.githubusercontent.com/14217/69177635-ea186300-0ad5-11ea-9aee-c54df87e9bc4.png)
- in editor settings, enable environment query system
![image](https://user-images.githubusercontent.com/14217/69177732-1a600180-0ad6-11ea-93a8-8923bf63ed10.png)
- create `EQS_FindSpawnLocation`
![image](https://user-images.githubusercontent.com/14217/69177875-6ad75f00-0ad6-11ea-8757-f69b3169bf49.png)
- create new blueprint class `EnvQueryContext_BotSpawns`  deriving from env query context blueprint
![image](https://user-images.githubusercontent.com/14217/69177996-b12cbe00-0ad6-11ea-9c6e-e1255fa24c35.png)
  - override function `ProvideActorsSet` and use `TargetPoint` for now
![image](https://user-images.githubusercontent.com/14217/69178215-25676180-0ad7-11ea-8b4b-1d1eef4c61fe.png)
-  in `EQS_FindSpawnLocaation` update `Generate Around` Parameter to the env query context
![image](https://user-images.githubusercontent.com/14217/69178338-63648580-0ad7-11ea-999f-96a111a42cf7.png)
- to visualize whats going on, create a new bp class deriving from EQS testing pawn `BP_EQSTestingPawn`
![image](https://user-images.githubusercontent.com/14217/69178421-8b53e900-0ad7-11ea-87cf-48c797326e5b.png)
  - add to level and update relevant eqs property on the instance
![image](https://user-images.githubusercontent.com/14217/69178536-d4a43880-0ad7-11ea-9249-8a535c3f7f7f.png)
  - add targetpoints to level and notice grid appearing when selecting the targetpawn
![image](https://user-images.githubusercontent.com/14217/69178831-6449e700-0ad8-11ea-896c-555c6af0e677.png)
  - we can use targetpoints to create spawn areas
- update `BP_TestGameMode`
![image](https://user-images.githubusercontent.com/14217/69178962-a6732880-0ad8-11ea-8cfe-9be1139df0ad.png)
- create distance test in `EQS_FindSpawnLocation`, note that querier is game mode (not so useful)
![image](https://user-images.githubusercontent.com/14217/69179025-c60a5100-0ad8-11ea-922d-7fefce17d699.png)
- create blueprint class `EnvQueryContext_AllPlayers` deriving from env query context blueprint
  - only return players that are alive
![image](https://user-images.githubusercontent.com/14217/69180142-cd325e80-0ada-11ea-92bd-dfc4e23a403e.png)
- modify `EQS_FindSpawnLocation` distance to get distance from all players
![image](https://user-images.githubusercontent.com/14217/69180308-0f5ba000-0adb-11ea-9a16-facb32075ca8.png)
- add player pawn to level and select testing pawn
![image](https://user-images.githubusercontent.com/14217/69180403-3f0aa800-0adb-11ea-8e05-a5ca676b46a8.png)
- modify eqs logic again, set distance minimum to 200 and observe score function
![image](https://user-images.githubusercontent.com/14217/69180560-7e38f900-0adb-11ea-9298-190aae22addb.png)
- observe results of score function around the player pawn
![image](https://user-images.githubusercontent.com/14217/69180670-aaed1080-0adb-11ea-9a9d-fe3704bdd59c.png)
  - we're going to use this logic to determine bot spawn locations, so min distance should probably be ~ 2000
